### PR TITLE
Added accountGetUserCount and accountListAllUsers

### PR DESCRIPTION
**Pull request by 3Daniel**

### CODE REVIEW

File: nx/include/switch/services/acc.h [hunk 1]
```diff
@@ -33,6 +35,15 @@ Result accountInitialize(void);
 void accountExit(void);
 Service* accountGetService(void);
 
+/// Get the total number of user profiles
+Result accountGetUserCount(s32* user_count);
+
+/**
+ * @brief Get a list of all user IDs
+ * @param userIDs Pointer to array of user IDs, array must be exactly size 8
```
Comment (yellows8) [inline nx/include/switch/services/acc.h]: Replace 8 with ACC_USER_LIST_SIZE.

Does official user-process sw have this size hard-coded to this?

**Comment by yellows8** [comment]
Change it to ```static Result _accountListAllUsers```, and impl a wrapper func like this: ```Result _accountListAllUsers(u128* userIDs, size_t max_userIDs, size_t *actual_total);```

Where actual_total is calculated by how many userIDs are non-zero, stopping at the first all-zero one.

**Comment by jyapayne** [comment]
@3Daniel Are you going to update this? I'm super looking forward to being able to use this in my homebrew, instead of asking the user to open a game first (which doesn't really work on hekate).

**Comment by 3Daniel** [comment]
@yellows8 can I please get a little clarification before I update this, just to make sure I understand you correctly.

I think it would be easier to use the `accountGetUserCount(s32* user_count)` function inside the wrapper to get the `actual_total` rather than calculate it.

I'm not clear on the purpose of `max_userIDs`. Is it so that the caller can specify a different maximum size for the `userIDs` array other than 8? ie. if `max_userIDs` is 3, then the `userIDs` array size is 3 and only the first 3 user IDs are returned?

**Comment by yellows8** [comment]
accountGetUserCount is unnecessary, it would be just a simple for-loop. "Is it so that the caller can specify a different maximum size for the userIDs array other than 8?" Yes.

File: nx/source/services/acc.c [hunk 2]
```diff
@@ -108,6 +109,38 @@ Result accountListAllUsers(u128* userIDs)
     return rc;
 }
 
+Result accountListAllUsers(u128* userIDs, size_t max_userIDs, size_t *actual_total)
+{
+    Result rc=0;
+
+    u128 *temp_userIDs;
+	temp_userIDs = malloc(sizeof(u128) * ACC_USER_LIST_SIZE);
```
Comment (fincs) [inline nx/source/services/acc.c]: Tab.

File: nx/source/services/acc.c [hunk 3]
```diff
@@ -33,6 +34,113 @@ Service* accountGetService(void) {
...
+
+    rc = _accountListAllUsers(temp_userIDs);
+
+    if (R_SUCCEEDED(rc)) {
+        size_t total_userIDs = 0;
+
+        for (int i = 0; i < ACC_USER_LIST_SIZE; i++) {
+            if (temp_userIDs[i]) {
```
Comment (fincs) [inline nx/source/services/acc.c]: Should be the other way around, to match what yls8 requested: `if (!temp_userIDs[i]) break;`

File: nx/source/services/acc.c [hunk 4]
```diff
@@ -33,6 +34,113 @@ Service* accountGetService(void) {
...
+
+        if (max_userIDs > total_userIDs) {
+            max_userIDs = total_userIDs;
+        }
+
+        for (int i = 0; i < max_userIDs; i++) {
+            userIDs[i] = temp_userIDs[i];
+        }
+
+        *actual_total = total_userIDs;
```
Comment (fincs) [inline nx/source/services/acc.c]: This is cumbersome. *actual_total should be assigned max_userIDs, which is the number of users that were returned.

Comment (3Daniel) [reply nx/source/services/acc.c]: Happy to change it, but this contradicts the original comment which said:
"Where actual_total is calculated by how many userIDs are non-zero, stopping at the first all-zero one."
The number which are non-zero is likely to be different to the max_userIDs.
Please confirm which way you'd prefer.

Comment (3Daniel) [reply nx/source/services/acc.c]: Yeah, that makes sense. I've changed it to be as you suggested now.

File: nx/source/services/acc.c (hunk 3 above)
Comment (3Daniel) [reply nx/source/services/acc.c]: That is a lot cleaner, thank you for the suggestion :-)

File: nx/source/services/acc.c (hunk 2 above)
Comment (3Daniel) [reply nx/source/services/acc.c]: Oops... that was silly of me, I think it should be fixed now.

File: nx/source/services/acc.c (hunk 4 above)
Comment (fincs) [reply nx/source/services/acc.c]: The original statement is sort of ambiguous, but it's a common API convention to do it like what I said.

File: nx/include/switch/services/acc.h [hunk 5]
```diff
@@ -33,6 +35,17 @@ Result accountInitialize(void);
 void accountExit(void);
 Service* accountGetService(void);
 
+/// Get the total number of user profiles
+Result accountGetUserCount(s32* user_count);
+
+/**
+ * @brief Get a list of all user IDs. The returned list will never be larger than ACC_USER_LIST_SIZE
```
Comment (yellows8) [inline nx/include/switch/services/acc.h]: Periods missing.

File: nx/source/services/acc.c [hunk 6]
```diff
@@ -33,6 +34,113 @@ Service* accountGetService(void) {
...
+    }
+
+    return rc;
+}
+
+Result accountListAllUsers(u128* userIDs, size_t max_userIDs, size_t *actual_total)
+{
+    Result rc=0;
+
+    u128 *temp_userIDs;
```
Comment (yellows8) [inline nx/source/services/acc.c]: Move to stack.

File: nx/source/services/acc.c [hunk 7]
```diff
@@ -33,6 +34,113 @@ Service* accountGetService(void) {
...
+        for (total_userIDs = 0; total_userIDs < ACC_USER_LIST_SIZE; total_userIDs++) {
+            if (!temp_userIDs[total_userIDs])
+                break;
+        }
+
+        if (max_userIDs > total_userIDs) {
+            max_userIDs = total_userIDs;
+        }
+
+        for (int i = 0; i < max_userIDs; i++) {
```
Comment (yellows8) [inline nx/source/services/acc.c]: memcpy

File: nx/source/services/acc.c [hunk 8]
```diff
@@ -112,8 +112,7 @@ static Result _accountListAllUsers(u128* userIDs)
 Result accountListAllUsers(u128* userIDs, size_t max_userIDs, size_t *actual_total)
 {
     Result rc=0;
-    u128 *temp_userIDs;
-    temp_userIDs = malloc(sizeof(u128) * ACC_USER_LIST_SIZE);
+    u128 *temp_userIDs = malloc(sizeof(u128) * ACC_USER_LIST_SIZE);
```
Comment (fincs) [inline nx/source/services/acc.c]: u128 temp_userIDs[ACC_USER_LIST_SIZE];

File: nx/source/services/acc.c [hunk 9]
```diff
@@ -33,6 +34,108 @@ Service* accountGetService(void) {
...
+
+        if (max_userIDs > total_userIDs) {
+            max_userIDs = total_userIDs;
+        }
+
+        memcpy(userIDs, temp_userIDs, sizeof(u128) * max_userIDs);
+        *actual_total = max_userIDs;
+    }
+
+    free(temp_userIDs);
```
Comment (fincs) [inline nx/source/services/acc.c]: Remove this due to above being changed to stack.

**Comment by fincs** [comment]
Any updates regarding the above comments? We'd like to get PRs merged in order to facilitate a release.

**Comment by 3Daniel** [comment]
Apologies for the delay. I have updated it to reflect your comments. I really appreciate you taking the time to provide feedback, even on small things. Thanks :-)

File: nx/source/services/acc.c [hunk 10]
```diff
@@ -33,6 +34,106 @@ Service* accountGetService(void) {
...
+    }
+
+    return rc;
+}
+
+Result accountListAllUsers(u128* userIDs, size_t max_userIDs, size_t *actual_total)
+{
+    Result rc=0;
+    u128 temp_userIDs[ACC_USER_LIST_SIZE];
+
```
Comment (yellows8) [inline nx/source/services/acc.c]: memset(temp_userIDs, 0, sizeof(temp_userIDs));

Comment (fincs) [reply nx/source/services/acc.c]: Or just " = {0};` at the end of the declaration I guess.

File: nx/source/services/acc.c [hunk 11]
```diff
@@ -1,4 +1,5 @@
 #include <string.h>
+#include <malloc.h>
```
Comment (yellows8) [inline nx/source/services/acc.c]: No longer needed.

**Pull request merged**